cognite/client/utils/_time.py

```diff
@@ -173,6 +395,152 @@ def split_time_range(start: int, end: int, n_splits: int, granularity_in_ms: int
+
+    freq = multiplier * GRANULARITY_IN_HOURS[unit]
+    index = pandas_date_range_tz(start, end, f"{freq}H")
+    expected_freq = pd.Timedelta(hours=freq)
+    next_mask = (index - index.to_series().shift(1)) != expected_freq
+    last_mask = (index.to_series().shift(-1) - index) != expected_freq
+    index = index[last_mask | next_mask]
+
+    hour, zero = pd.Timedelta(hours=1), pd.Timedelta(0)
+    transitions = []
+    for start, end in zip(index[:-1], index[1:]):
+        if start.dst() == end.dst():
+            dst_adjustment = 0
+        elif start.dst() == hour and end.dst() == zero:
+            # Fall, going away from summer.
+            dst_adjustment = 1
+        elif start.dst() == zero and end.dst() == hour:
+            # Spring, going to summer.
+            dst_adjustment = -1
+        else:
+            raise ValueError(f"Invalid dst, {start} and {end}")
+
+        transitions.append(
+            {
+                "start": start.to_pydatetime().astimezone(utc),  # type: ignore
+                "end": end.to_pydatetime().astimezone(utc),  # type: ignore
+                "granularity": f"{freq+dst_adjustment}h",
+            }
+        )
+    return transitions
+
+
+def pandas_date_range_tz(start: datetime, end: datetime, freq: str, inclusive: str = "both") -> pandas.DatetimeIndex:
+    """
+    Pandas date_range struggles with time zone aware datetimes.
+    This function overcomes that limitation.
+    """
+    pd = local_import("pandas")
+    # Pandas seems to have issues with ZoneInfo object, so removing the timezone and adding it back.
+    return pd.date_range(start.replace(tzinfo=None), end.replace(tzinfo=None), freq=freq, inclusive=inclusive).tz_localize(start.tzinfo.key)  # type: ignore [union-attr]
+
+
+def validate_timezone(start: datetime, end: datetime) -> ZoneInfo:
+    if sys.version_info >= (3, 9):
+        from zoneinfo import ZoneInfo
+    else:
+        from backports.zoneinfo import ZoneInfo
+
+    if missing := [name for name, timestamp in zip(("start", "end"), (start, end)) if not timestamp.tzinfo]:
+        names = " and ".join(missing)
+        end_sentence = " do not have timezones." if len(missing) >= 2 else " does not have a timezone."
+        raise ValueError(f"All times must be time zone aware, {names}{end_sentence}")
+
+    if not isinstance(start.tzinfo, ZoneInfo) or not isinstance(end.tzinfo, ZoneInfo):
+        raise ValueError("Only ZoneInfo implementation of tzinfo is supported")
+
+    if start.tzinfo is not end.tzinfo:
+        raise ValueError(f"start and end have different timezones, {start.tzinfo.key!r} and {end.tzinfo.key!r}.")
+
+    return start.tzinfo
+
+
+def to_pandas_freq(granularity: str, start: datetime) -> str:
+    multiplier, unit = get_granularity_multiplier_and_unit(granularity, standardize=True)
+
+    unit = {"s": "S", "m": "T", "h": "H", "d": "D", "w": "W-MON", "month": "MS", "quarter": "QS", "year": "AS"}.get(
+        unit, unit
+    )
```

Comment:
Awesome!! This would allow us to more easily add week days later! 🚀

Reply:
Should it be applied in the old `retrieve_dataframe `as well? There there is a `.replace("m", "T") `to achieve the same.

Reply:
Let's add that separately/later. The function will reject any "non-valid" granularities currently and I think it would be nice for us to add week/month/quarter/year